File: File Templates/Remote Push Notifications/Push Notification Center OBJC.xctemplate/___FILEBASENAME___.m
//___FILEHEADER___

#import "___FILEBASENAME___.h"
#import <UIKit/UIKit.h>
#import <UserNotifications/UserNotifications.h>

@interface ___FILEBASENAMEASIDENTIFIER___ () <UNUserNotificationCenterDelegate>

@property (nonatomic, strong) UNUserNotificationCenter *notificationCenter;

@end

@implementation ___FILEBASENAMEASIDENTIFIER___

- (instancetype)init {
    self = [super init];
    if (self) {
        _notificationCenter = [UNUserNotificationCenter currentNotificationCenter];
        _notificationCenter.delegate = self;
    }
    return self;
}

- (void)registerForRemoteNotification {
    UNAuthorizationOptions authOptions = UNAuthorizationOptionBadge + UNAuthorizationOptionSound + UNAuthorizationOptionAlert;
    
    __weak typeof(self) weakSelf = self;
    
    [self.notificationCenter requestAuthorizationWithOptions:authOptions completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if (granted) {
            [weakSelf.notificationCenter getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
                if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [[UIApplication sharedApplication] registerForRemoteNotifications];
                    });
                }
            }];
        }
    }];
}

- (void)handleRemoteNotification:(NSDictionary *)notification {
    
}

- (void)didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {

}

#pragma mark - UNUserNotificationCenterDelegate -

- (void)userNotificationCenter:(UNUserNotificationCenter *)center
       willPresentNotification:(UNNotification *)notification
         withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler {
    
}

@end

File: File Templates/Remote Push Notifications/Push Notification Center SWIFT.xctemplate/___FILEBASENAME___.swift
//___FILEHEADER___

import Foundation
import UIKit
import UserNotifications

class ___FILEBASENAMEASIDENTIFIER___: NSObject {
    
    private func configure() {
        UNUserNotificationCenter.current().delegate = self
    }

    public func registerForRemoteNotifications() {
        configure()
        
        let application = UIApplication.shared
        let authOptions: UNAuthorizationOptions = [.alert, .badge, .sound]
        UNUserNotificationCenter.current().requestAuthorization(options: authOptions) {_, _ in

        }

        application.registerForRemoteNotifications()
        
    }

    public func handleRemoteNotification(with userInfo: [AnyHashable: Any]) {
        
    }
    
    public func handleDeviceToken(token: Data) {
        let strToken = token.reduce("") { $0 + String(format: "%02.2hhx", $1) }
    }
}

extension ___FILEBASENAMEASIDENTIFIER___: UNUserNotificationCenterDelegate {

    func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        
    }
}
